e2e/nomostest/registry_server.go

```diff
@@ -97,32 +98,136 @@ func setupRegistry(nt *NT) error {
 	return nil
 }
 
-// setupRegistryClient handles registry authentication.
-// This is necessary on each test because gcloud auth tokens expire after 1h and
-// some test suites take longer than 1h.
+// setupRegistryClient handles registry authentication and port-forwarding.
+//
+// Re-auth is necessary on each test because gcloud auth tokens expire after 1h
+// and some test suites take longer than 1h.
+//
+// Port-forwarding is configured for local registries so that if the registry
+// pod crashes or is rescheduled the port-forwarding will be restarted after it
+// becomes healthy again. Then the client will re-login and re-push all the
+// images and charts that were previously pushed during the same test.
+//
+// For non-local registries, only login, logoiut, and cleanup are performed.
+//
+// Helm client is only configured if RequireHelmProvider is true.
+// OCI client is only configured if RequireOCIProvider is true.
 func setupRegistryClient(nt *NT, opts *ntopts.New) error {
-	nt.T.Cleanup(func() {
-		if opts.RequireHelmProvider || opts.RequireLocalHelmProvider {
-			if err := nt.HelmProvider.Logout(); err != nil {
-				nt.T.Error(err)
+
+	// For local registries, set up port-forwarding to the in-cluster registry server.
+	if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local || opts.RequireHelmProvider && *e2e.HelmProvider == e2e.Local {
+		// The local registry uses ephemeral storage. So login and re-push all
+		// the oci images and helm charts after the registry recovers from a crash.
+		// Before the test, nt.ociImages & nt.helmPackages will be empty.
+		var onReadyCallbacks []func(int, string)
+		if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local {
```

Comment:
nit: we can check either `opts.RequireOCIProvider` or `*e2e.OCIProvider == e2e.Local`. It's redundant to check both as line 118 guarantees the other one.

Reply:
Not true. It's possible for the provider to be local but the current test not to require OCI, if the current test requires Helm, for example.